### ISSUE TITLE
freeze the 3D ResNet, use misalignment score as loss function and optimize 2D shifts

### ISSUE THREAD

**Issue by alisterburt** [warpem/miss-alignment#3]
Attempt to back-propagate from the alignment score through the network and reconstruction algorithm to the 2D shifts of each projected image to get a gradient to iteratively optimize the shifts to minimize the misalignment score

Create an optimizer for the parameters defining the 2D shifts then
- Generate tilt series same as in #6 with 2D shifts as parameters
- apply 2D shifts then perform 3D reconstruction
- Feed the resulting 3D reconstruction through the frozen ResNet.
- Take score maximization as the loss.
- Back-propagate all the way back to the 2D shifts.
- Advance optimization.
- 
Repeat until convergence.
End up with a perfectly aligned reconstruction?

**Comment by McHaillet** [comment]
When further testing the alignment optimization, I noticed some weird rotation issues and upon debugging found we were applying the random rotation for each volume in the wrong order. We did tilt_rotation_matrices @ random_rotation, but should have done random_rotation @ tilt_rotation_matrices. This was leading to weird reconstructions and part of the reason why the alignment was not fully working.

I partly trained a model today (its not finished, but please see the .ckpt file i shared from ~ epoch 90) and it works pretty well for reducing misalignment. Model was trained on branch `test_loss_regu`, commit 50ed655 -- train data inflated by factor 60 and with this command

```bash
miss-alignment train --dataset-directory datasets/emdb/train/ --output-directory experiments/test006 --batch-size 64 --epochs 310 --n-workers 16 --learning-rate 1e-03 --seed 65123134
```

To emphasize this model still uses the contrastive marginrank loss as this bug was actually preventing me from seeing that the model was able to correct for the trajectory based translations.

With this latest version you can also run this on the test dataset to align some examples from the zenodo archive:

```bash
miss-alignment optimize_alignment --model-checkpoint [insert checkpoint] --test-data-directory emdb/test/ --output-directory align_test/
```

 Thought you guys might be excited to give it a go!

For next week I need to:
* ensure global shift does not drift away
* do numerical evaluation of detected shifts

**Comment by McHaillet** [comment]
Hey! Still some work on branch `test_loss_regu` commit fa9472f

I did the evaluation on some larger batches to confirm that the model corrects y-shifts better than x-shifts. The model was trained with the standard margin rank loss from torch with pairs.

![Image](https://github.com/user-attachments/assets/1e71ec43-bc52-457e-99a0-6fb1410c6769)

_Points in the graphs are the sum of the image shifts for one tilt-series before (blue) and after (orange) correction. The lower the sum the less misalignment. I plotted it separately along x and y. (Note to self: use dot instead of line plot + force y axis limit)_

I trained a second model with margin rank and an l2-norm but it performs **margin**ally worse.

![Image](https://github.com/user-attachments/assets/4cfba134-7379-4794-8384-17989d1862f7)


I am still curious to try out the triplet loss.

**Comment by dtegunov** [comment]
For comparison, what did the misalignment profile of the "aligned" target (the one used as positive example) look like? Have you tried training on X jitter only to make it not pay attention to Y?

It could be useful to benchmark the alignment accuracy on exactly the same misalignment of the same maps in all cases. Just mentioning this because the misaligned profiles look different between the two rows of plots.

**Comment by McHaillet** [comment]
> It could be useful to benchmark the alignment accuracy on exactly the same misalignment of the same maps in all cases. Just mentioning this because the misaligned profiles look different between the two rows of plots.

Indeed they were. These are the same plots but with seeding

marginrankingloss

![Image](https://github.com/user-attachments/assets/0cedde0d-f716-47fa-83ff-0bce380e02fc)

marginrankingloss + l2norm

![Image](https://github.com/user-attachments/assets/8c515351-96a8-447a-8e0b-6bba24d6af07)

With l2 norm is definitely worse. But, you could still experiments with strength of l2 norm and the margin distance -- probably the relation between the two makes the model worse. I think that parameter space is not worth it to explore.

**Comment by McHaillet** [comment]
This is intermediate result with model that also sees examples with shifts in a single direction:

![Image](https://github.com/user-attachments/assets/9610fe7e-2c55-48d0-afb5-ccea042544df)

Appears to perform better in x-shifts. I'll add an update tomorrow, once it finished training.

**Comment by alisterburt** [comment]
Am I interpreting it right that it's still performing quite a bit worse on x shifts than y shifts? (Even if better than before?)

Thanks for sharing!

**Comment by McHaillet** [comment]
> Am I interpreting it right that it's still performing quite a bit worse on x shifts than y shifts? (Even if better than before?)

You're seeing that correctly!

**Comment by McHaillet** [comment]
After a full training run, it might be slightly better than the model trained without single direction shifts, but definitely still performing worse in x than in y

![Image](https://github.com/user-attachments/assets/3b1b4af3-f7cd-4b2b-92dd-460629798762)

**Comment by dtegunov** [comment]
> This is intermediate result with model that also sees examples with shifts in a single direction:
> 
> ![Image](https://github.com/user-attachments/assets/9610fe7e-2c55-48d0-afb5-ccea042544df)
> 
> Appears to perform better in x-shifts. I'll add an update tomorrow, once it finished training.

Do I understand correctly the model sees a mixture of examples with xy, x, and y shifts? It might be interesting to see how the losses between these example groups differ.

**Comment by dtegunov** [comment]
Does the optimization work better if you take gradients averaged over multiple volumes misaligned in the same way?

**Comment by McHaillet** [comment]
> Does the optimization work better if you take gradients averaged over multiple volumes misaligned in the same way?

Just tried that, it absolutely does (this is with averaging over 4 boxes):

![Image](https://github.com/user-attachments/assets/2f282b1d-cfe0-42be-94d7-79f376ab651e)

*edit
and with 8 boxes (sorry couldn't keep the seeding constant -- not sure why) -> looks even better though imo

![Image](https://github.com/user-attachments/assets/e1eff878-74e2-4c93-9f8a-f6551611a3fb)

**Comment by dtegunov** [comment]
Yay! Now the question is if it's because it has more voxels to look at, or because there are some maps in the mix that it works better with. You could test e.g. for each map and a set of random orientations and misalignments how well it aligns on average. There might be interesting correlations beyond the likely "bigger map = better alignment".

Oh, and there might also be correlation between how well Miss does vs. conventional cross-correlation between tilts.

**Comment by McHaillet** [comment]
Done ✅
